autode/bracket/ieip.py

```diff
@@ -0,0 +1,626 @@
+        for coord in [self.left_coords, self.right_coords]:
+            parall_g = d_hat * np.dot(d_hat, coord.g)
+            perp_g = coord.g - parall_g
+            rms_perp_g = np.sqrt(np.mean(np.square(perp_g)))
+            perp_gradients.append(rms_perp_g)
+
+        return perp_gradients
+
+    def redistribute_imagepair(
+        self,
+        ll_neb_interp: bool = True,
+        interp_fraction: float = 1 / 4,
+    ):
+        """
+        Redistribute the image pair by running a NEB calculation at lmethod or
+        use only IDPP interpolation, and then fitting a cubic spline on energy
+        calculated by the method (low_sp keywords). It generates the image pair
+        on both sides of the peak on the fitted spline, with a distance of
+        interp_fraction * total path distance on either side.
+
+        Args:
+            ll_neb_interp (bool): Whether to optimise the interpolated path with
+                                  NEB at lmethod for the interpolation.
+
+            interp_fraction (float): Fraction of total interpolated path distance
+                                     that will be used to generate the image pair
+                                     on either side of the interpolated TS
+        """
+        # Use at least 5 images for interpolation
+        n_images = int(_interp_image_density * self.dist - 1)
+        n_images = max(n_images, 5 + 2)
+
+        interp = NEB.from_end_points(
+            self._left_image.copy(), self._right_image.copy(), n_images
+        )
+        if ll_neb_interp:
+            interp.calculate(method=get_lmethod(), n_cores=self._n_cores)
+
+        # Only calc intermediate images, initial and final already have energies
+        path_points = interp.images[1:-1]
+        # Get energies
+        assert self.left_coords.e and self.right_coords.e
+        path_energies = _parallel_calc_energies(
+            path_points, method=self._method, n_cores=self._n_cores
+        )
+        energies = [self.left_coords.e] + path_energies + [self.right_coords.e]
+        logger.info(
+            f"Fitting parametric spline on {len(interp.images)} points"
+        )
+
+        # NOTE: Here we are fitting a parametric spline, with the parameter
+        # being the path length along approx. rxn coordinate and target being all
+        # coordinates *and* energy at the points
+        path_spline = CubicPathSpline.from_species_list(interp.images)
+        path_spline.fit_energies(energies)
+        peak_x = path_spline.energy_peak()
+        if peak_x is None:
+            raise RuntimeError(
+                "The fitted spline does not have a peak! Unable to proceed"
+            )
+        # Check the peak is not at the beginning or end
+        assert 0.01 < peak_x < 0.99
+        # convert to integrated arc lengths
+        peak_pos = path_spline.path_integral(0, peak_x)
+        path_length = path_spline.path_integral(0, 1)
+
+        # todo double check spline code
```

Comment:
🎏

Reply:
@shoubhikraj did you want to check this?

Reply:
@t-young31 Yes haven't managed to check it yet. I will run some extra reactions to see if the interpolation is reasonable enough. that's why i kept this open.

Reply:
@t-young31 Ok, I ran some checks. Everything seems to be running fine.